mediatek: add support for WAVLINK WL-WN536AX6 Rev a

ses1er opened the pull request:
Product name: Wavlink WL-WN536AX6 Rev a "Mighty LX2"
Product link: https://www.wavlink.com/en_us/product/WL-WN536AX6.html

Specifications:
SOC: MT7986AV
RAM: 512MB DDR4
Flash: 128MB SPI NAND
Ports: 4 LAN (1G) & 1 WAN (2.5G)
WIFI: MT7976PN + MT7975N
LEDs: 8 (STATUS, WIFI, WAN, LAN1, LAN2, LAN3, LAN4, PWR)
USB: 1 (3.0)

MAC table, same as stock firmware:
LAN: 80:3F:5D:xx:xx:x1 partition "hw" at 0x44e (ASCII)
WAN: 80:3F:5D:xx:xx:x2 partition "hw" at 0x460 (ASCII)
2G: 80:3F:5D:xx:xx:x3
5G: 80:3F:5D:xx:xx:x4

* Installation with OEM WebUI:

Note: Make sure PC is connected on LAN1 port. The OEM firmware has an unknown root password and settings are kept after upgrading firmware. Therefore, a customized Openwrt firmware is needed to remove the root password on startup, by adding `passwd -d root` to /etc/init.d/bootcount. The WebUI does a filename check so the customized firmware is named accordingly.

1. Download modified firmware file `WAVLINK_WN536AX6-A_M36AX6_V250320-WO-437baca-modified.bin` from https://github.com/ses1er/firmware-misc/tree/main/wavlink/wl-wn536ax6a
2. Log into WebUI on default IP: http://192.168.20.1
3. Browse to More (top menu) -> System -> Firmware Upgrade.
4. Under `Local Upgrade` section, check the device to be upgraded and upload downloaded modified firmware. Click `UPLOAD FILE`, then `APPLY`
5. Wait about 2 minutes (ignore progress bar), and browse to http://192.168.20.1. You should see LUCI login page. Username is root and no password.
6. Browse to `System -> Backup/Flash Firmware`, click on `Flash Image`, click `Browse` and locate `openwrt-mediatek-filogic-wavlink_wl-wn536ax6a-squashfs-sysupgrade.bin` file.
7. Uncheck `Keep settings and retain the current configuration` and click `Continue`.
8.  Router will now be set to IP 192.168.1.1 which is the Openwrt default.

* Installation with UART:

Note: Having UART connected while cold booting the device will result in a kernel panic when initializing wifi. I've found this workaround:
1. Power off the device and ensure UART is not connected to PC.
2. Power up the device, when lights come on, plug in UART.
3. Warm boots and soft restarts will not cause kernel panic for the duration of device being powered on. Repeat steps for subsequent cold boots.

Installation:
1. Configure TFTP server with IP 192.168.1.66. Copy `openwrt-mediatek-filogic-wavlink_wl-wn536ax6-a-initramfs.itb` to TFTP root.
2. Interrupt boot by pressing 0.
3. Run the following in Uboot console: `setenv serverip 192.168.1.66; setenv ipaddr 192.168.1.1; tftpboot 0x46000000 openwrt-mediatek-filogic-wavlink_wl-wn536ax6-a-initramfs.itb; bootm`
4. Transfer `openwrt-mediatek-filogic-wavlink_wl-wn536ax6-a-squashfs-sysupgrade.bin` to device: (`scp -O openwrt-mediatek-filogic-wavlink_wl-wn536ax6a-squashfs-sysupgrade.bin root@192.168.1.1:/tmp/`)
5. Run the following on device: `sysupgrade -n /tmp/openwrt-mediatek-filogic-wavlink_wl-wn536ax6-a-squashfs-sysupgrade.bin`

OEM firmware's DTS file: https://gist.github.com/ses1er/51b6cbd59609f3f733ef6efe01cc39e6

### Limitation

Since this device has 1 WLAN LED and 2 radios, you'd need to pick which radio controls the LED.
With #19903, either/both radios can activate the LED and provides an OEM experience to the device.

musashino205 commented on target/linux/mediatek/dts/mt7986a-wavlink-wl-wn536ax6a.dts:
```diff
@@ -0,0 +1,393 @@
+			color = <LED_COLOR_ID_RED>;
+			gpios = <&pio 15 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "none";
+		};
```
no `function`?

ses1er replied:
There is actually no red led on the device. It was on the original firmware's dts file. Should I remove this?

ses1er replied:
Actually, i'll remove `linux,default-trigger = "none";` portion since that was not part of the original firmware's dts file. Just in case there is a red led somewhere on the device.

ses1er replied:
FYI, OEM firmware's DTS file: https://gist.github.com/ses1er/51b6cbd59609f3f733ef6efe01cc39e6

achterin replied:
sometimes vendors just use the same base dts file for multiple devices, even if these devices do miss some features. so i'd say remove the red-led definition if there is no red led present.

ses1er replied:
I've removed the led. In addition, I've removed a non-existent button (touch).

musashino205 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -2407,6 +2407,24 @@ define Device/unielec_u7981-01-nand
 
+define Device/wavlink_wl-wn536ax6a
+  DEVICE_VENDOR := WAVLINK
+  DEVICE_MODEL := WL-WN536AX6 Rev a
```
```suggestion
  DEVICE_MODEL := WL-WN536AX6
  DEVICE_VARIANT := Rev a
```

musashino205 commented on target/linux/mediatek/dts/mt7986a-wavlink-wl-wn536ax6a.dts:
```diff
@@ -0,0 +1,393 @@
+			full-duplex;
+			link-gpio = <&pio 47 GPIO_ACTIVE_HIGH>;
+			pause;
+			phy-handle = <&phy5>;
```
This `phy-handle` is invalid.

musashino205 commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -229,6 +229,10 @@ tplink,re6000xd)
 	ucidef_set_led_netdev "lan-2" "lan-2" "blue:lan-1" "lan2" "link tx rx"
 	ucidef_set_led_netdev "eth1" "lan-3" "blue:lan-2" "eth1" "link tx rx"
 	;;
+wavlink,wl-wn536ax6a)
```
```suggestion
wavlink,wl-wn536ax6-a)
```

musashino205 commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -67,6 +67,7 @@ mediatek_setup_interfaces()
 	nradio,c8-668gl|\
 	ruijie,rg-x60-pro|\
 	unielec,u7981-01*|\
+	wavlink,wl-wn536ax6a|\
```
```suggestion
	wavlink,wl-wn536ax6-a|\
```

ses1er commented on target/linux/mediatek/dts/mt7986a-wavlink-wl-wn536ax6a.dts:
```diff
@@ -0,0 +1,393 @@
+#include "mt7986a.dtsi"
+
+/ {
+	compatible = "wavlink,wl-wn536ax6a", "mediatek,mt7986a";
```
fixed, thank you

ses1er commented on target/linux/mediatek/dts/mt7986a-wavlink-wl-wn536ax6-a.dts:
```diff
@@ -0,0 +1,386 @@
+
+/ {
+	compatible = "wavlink,wl-wn536ax6-a", "mediatek,mt7986a";
+	model = "WAVLINK WL-WN536AX6 Rev a";
```
The sticker on the bottom of the device shows "Rev a": https://openwrt.org/_media/media/wavlink/wl-wn536axa-bottom.jpg

achterin commented on target/linux/mediatek/dts/mt7986a-wavlink-wl-wn536ax6-a.dts:
```diff
@@ -0,0 +1,386 @@
+		led-boot = &led_status_blue;
+		led-failsafe = &led_status_blue;
+		led-running = &led_status_blue;
+		led-upgrade = &led_status_blue;
```
in that case, `&wifi` should be the correct device.

xdarklight commented on target/linux/mediatek/dts/mt7986a-wavlink-wl-wn536ax6-a.dts:
```diff
@@ -0,0 +1,376 @@
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200n1 loglevel=8  \t\t\t\tearlycon=uart8250,mmio32,0x11002000";
```
what's `loglevel=8`?
and why ` \t\t\t\t`?

xdarklight commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -2407,6 +2407,25 @@ define Device/unielec_u7981-01-nand
+  KERNEL_INITRAMFS_SUFFIX := .itb
+  KERNEL_IN_UBI := 1
+  DEVICE_PACKAGES := kmod-usb3 kmod-mt7915e kmod-mt7986-firmware mt7986-wo-firmware
+  SUPPORTED_DEVICES += mediatek,mt7986a-2500wan-spim-snand-rfb
```
`mediatek,mt7986a-2500wan-spim-snand-rfb` seems generic - I suspect it's only used for the "modified" factory firmware image.
My understanding is that we can't build a factory image here as we need a hack to reset the root user password. Can't this simply be part of the code that builds the "modified" factory image?

ses1er replied:
This is needed going from original firmware to modified firmware as well as going to Openwrt since factory sysupgrade retains settings.

I'm still kind of new to this so pardon my ignorance, but how would I build an openwrt factory image with specific specific files/settings? That's if I'm understanding your question correctly.

xdarklight commented on target/linux/mediatek/dts/mt7986a-wavlink-wl-wn536ax6-a.dts:
```diff
@@ -0,0 +1,376 @@
+
+		phy5: phy@5 {
+			reg = <5>;
+			compatible = "ethernet-phy-ieee802.3-c45";
```
I suspect that this is the MaxLinear 2.5G PHY.
My suggestion is to add a comment in case they switch out that PHY for another one (not sure whether WAVLINK did that or would do that) and then people start wondering why the reset delays are causing them problems.

ses1er replied:
comment added on `phy@6`. in addition, added comments on each of the `&gmac*` to indicate `LAN`/`WAN`.

xdarklight replied:
Thanks for testing!

I think I now understand why the 2.5G PHY has a `phy_id` of `0x00000000`: it's a C45 PHY which means there may be multiple sub-devices with individual IDs (as far as I understand, first time I'm seeing this). So the sysfs output is expected.
It also explains why `mdio-tools` finds the PHY ID: it simply reads the registers (while Linux seems to default to `phy_id = 0` for C45 PHYs: https://elixir.bootlin.com/linux/v6.12.58/source/drivers/net/phy/phy_device.c#L1017)

Regarding the reset pin: if flipping the polarity doesn't do anything then it's most likely the wrong GPIO.
Most vendors are using a reference design and adapting it to their needs.
I found the following ones in OpenWrt:
```
$ git grep 'reset-gpios = <&pio ' target/linux/mediatek/dts/mt7986a-* | cut -d'&' -f2 | cut -d' ' -f1,2 | sort | uniq
pio 10
pio 13
pio 17
pio 18
pio 19
pio 25
pio 27
pio 47
pio 49
pio 5
pio 6
```
I've never bricked a board by poking random GPIOs - at the same time it's not impossible that this happens.

My WL-WN536AX6 is on it's way it should be here by the weekend.
I'm happy to try out this patch and/or experiment with the GPIOs.

PS: I also want to bring https://lore.kernel.org/netdev/20241004102413.5838-7-SkyLake.Huang@mediatek.com/ into mainline so we can correctly describe the LAN LEDs in the `.dts` as well.

ses1er replied:
I updated the wiki:

```
Pry the top cover plate around the edge carefully as to not snap anything.
Start with the corner areas and loosen the rear part first, then the sides.
The most difficult part is the LED portion on the front side.
With the rear and sides loosened, slide your hand from the rear, place your hand under the top cover and carefully pry the LED section apart.
You may need to use your fingers for leverage. It should snap open with a bit of force. You may need to wiggle the top cover from side to side a bit.
```

I hope that helps. I'd recommend getting UART set up first to get a full NAND backup with stock firmware. Then try the modified firmware  via stock webui upgrade.

xdarklight reviewed (changes requested):
Thank you for the PR.
I'll have to see if my order gets cancelled or not - meaning: I cannot test it myself right now.

ses1er commented on target/linux/mediatek/dts/mt7986a-wavlink-wl-wn536ax6-a.dts:
```diff
@@ -0,0 +1,376 @@
+		conf-pd {
+			pins = "SPI2_CLK", "SPI2_MOSI", "SPI2_MISO";
+			bias-pull-down = <MTK_PUPD_SET_R1R0_11>;
+			drive-strength = <MTK_DRIVE_8mA>;
```
There was a recent commit that is the opposite of what you've stated: https://github.com/openwrt/openwrt/commit/f3ce0c29ea0ec56b702a38f3aca84fb6648008ba

Are we going back?

xdarklight replied:
oh, that's surprising.

upstream Linux' `Documentation/devicetree/bindings/pinctrl/mediatek,mt7986-pinctrl.yaml` clearly states the allows `drive-strength` values and the tooling within the Linux kernel can validate `.dts{,i}` files against that schema.
In other words: `drive-strength = <4>;` would pass while `drive-strength = <123>;` would fail the validation.

I don't know why OpenWrt would want to take a different route.
That said, I consider that discussion not part of this PR.
For consistency reasons your approach is fine (and the resulting `.dtb` is the same anyways).

ses1er commented on target/linux/mediatek/dts/mt7986a-wavlink-wl-wn536ax6-a.dts:
```diff
@@ -0,0 +1,376 @@
+
+		fixed-link {
+			full-duplex;
+			link-gpio = <&pio 47 GPIO_ACTIVE_HIGH>;
```
removed `link-gpio` and tested.

xdarklight commented on target/linux/mediatek/dts/mt7986a-wavlink-wl-wn536ax6-a.dts:
```diff
@@ -0,0 +1,376 @@
+			compatible = "ethernet-phy-ieee802.3-c45";
+			reg = <6>;
+			phy-mode = "2500base-x";
+		};
```
Apologies if I previously caused confusion with this.
This description means (see `Documentation/devicetree/bindings/net/mdio.yaml` in the Linux kernel: 
> The phandle and specifier for the GPIO that controls the RESET lines of all devices on that MDIO bus.

This is obviously not correct because the switch (below) has it's own reset-gpio (`&pio 5` - which I have confirmed is correct by flipping it's polarity).

I understand now where all the problems were coming from with the original approach.
This is what I came up with:
```suggestion
		/* MaxLinear GPY211C 2.5G PHY */
		phy6: phy@6 {
			/*
			 * Force the ID here as the PHY only reports it's
			 * (C45) ID correctly after a reset (before resetting
			 * it reports 0xfffffff, which causes only the genphy
			 * driver to match).
			 */
			compatible = "ethernet-phy-id67c9.de10";
			reg = <6>;
			reset-gpios = <&pio 6 GPIO_ACTIVE_LOW>;
			reset-assert-us = <600>;
			reset-deassert-us = <20000>;
			phy-mode = "2500base-x";
		};
```

Using `compatible = "ethernet-phy-ieee802.3-c45"` results in: `mtk_soc_eth 15100000.ethernet eth1: PHY [mdio-bus:06] driver [Generic Clause 45 PHY] (irq=POLL)`
Forcing it's ID results in (just like with your approach to move the reset-gpios up): `mtk_soc_eth 15100000.ethernet eth1: PHY [mdio-bus:06] driver [Maxlinear Ethernet GPY211C] (irq=POLL)`

For this case `Documentation/devicetree/bindings/net/ethernet-phy.yaml` states:
>       - pattern: "^ethernet-phy-id[a-f0-9]{4}\\.[a-f0-9]{4}$"
>         description:
>           If the PHY reports an incorrect ID (or none at all) then the
>           compatible list may contain an entry with the correct PHY ID
>           in the above form.
>           The first group of digits is the 16 bit Phy Identifier 1
>           register, this is the chip vendor OUI bits 3:18. The
>           second group of digits is the Phy Identifier 2 register,
>           this is the chip vendor OUI bits 19:24, followed by 10
>           bits of a vendor specific ID.

In case you're wondering why genphy doesn't work (on some PHYs it will work but not provide full functionality): `mxl-gpy.c` has a custom `gpy_read_status` function which even has it's own handling for reading the link status.

Does this approach also work on your device?

xdarklight reviewed (commented):
First of all: thank you for your hard work!
The trick with prying with my hand to open the case worked, no clips were harmed in the process.

I have good and more good news:
- it works on my device!
- I was able to come up with (what I think is) the root cause of the PHY reset issues, see my comment specific to that

so far I have successfully tested:
- LAN ports
- WAN port
- USB
- LEDs

For users that want to backup their whole flash they need to add a custom partition that spans the whole 128MiB.
Then with `CONFIG_PACKAGE_nand-utils=y` they can: `nanddump --file <directory>/full.nanddump /dev/mtd6` (assuming `mtd6` the partition that spans the whole flash)

ses1er committed:
mediatek: add support for WAVLINK WL-WN536AX6 Rev a

Product name: Wavlink WL-WN536AX6 Rev a "Mighty LX2"
Product link: https://www.wavlink.com/en_us/product/WL-WN536AX6.html

Specifications:
SOC: MT7986AV
RAM: 512MB DDR4
Flash: 128MB SPI NAND
Ports: 4 LAN (1G) & 1 WAN (2.5G)
WIFI: MT7976PN + MT7975N
LEDs: 8 (STATUS, WIFI, WAN, LAN1, LAN2, LAN3, LAN4, PWR)
USB: 1 (3.0)

MAC table, same as stock firmware:
LAN: 80:3F:5D:xx:xx:x1 partition "hw" at 0x44e (ASCII)
WAN: 80:3F:5D:xx:xx:x2 partition "hw" at 0x460 (ASCII)
2G: 80:3F:5D:xx:xx:x3
5G: 80:3F:5D:xx:xx:x4

* Installation with OEM WebUI:

Note: Make sure PC is connected on LAN1 port. The OEM firmware has an unknown root password and settings are kept after upgrading firmware. Therefore, a customized Openwrt firmware is needed to remove the root password on login, by adding `passwd -d root` to /etc/init.d/bootcount. The WebUI does a filename check so the customized firmware is named accordingly.

1. Download modified firmware file `WAVLINK_WN536AX6-A_M36AX6_V250320-WO-437baca-modified.bin` from https://github.com/ses1er/firmware-misc/tree/main/wavlink/wl-wn536ax6a
2. Log into WebUI on default IP: http://192.168.20.1
3. Browse to More (top menu) -> System -> Firmware Upgrade.
4. Under `Local Upgrade` section, check the device to be upgraded and upload downloaded modified firmware. Click `UPLOAD FILE`, then `APPLY`
5. Wait about 2 minutes (ignore progress bar), and browse to http://192.168.20.1. You should see LUCI login page. Username is root and no password.
6. Browse to `System -> Backup/Flash Firmware`, click on `Flash Image`, click `Browse` and locate `openwrt-mediatek-filogic-wavlink_wl-wn536ax6-a-squashfs-sysupgrade.bin` file.
7. Uncheck `Keep settings and retain the current configuration` and click `Continue`.
8. Router will now be set to IP 192.168.1.1 which is the Openwrt default.

* Installation with UART:

Note: Having UART connected while cold booting the device will result in a kernel panic when initializing wifi. I've found this workaround:
1. Power off the device and ensure UART is not connected to PC.
2. Power up the device, when lights come on, plug in UART.
3. Warm boots and soft restarts will not cause kernel panic for the duration of device being powered on. Repeat steps for subsequent cold boots.

1. Configure TFTP server with IP 192.168.1.66. Copy `openwrt-mediatek-filogic-wavlink_wl-wn536ax6a-initramfs.itb` to TFTP root.
2. Interrupt boot by pressing 0.
3. Run the following in Uboot console: `setenv serverip 192.168.1.66; setenv ipaddr 192.168.1.1; tftpboot 0x46000000 openwrt-mediatek-filogic-wavlink_wl-wn536ax6-a-initramfs.itb; bootm`
4. Transfer `openwrt-mediatek-filogic-wavlink_wl-wn536ax6-a-squashfs-sysupgrade.bin` to device:
   (`scp -O openwrt-mediatek-filogic-wavlink_wl-wn536ax6-a-squashfs-sysupgrade.bin root@192.168.1.1:/tmp/`)
5. Run the following on device: `sysupgrade -n /tmp/openwrt-mediatek-filogic-wavlink_wl-wn536ax6-a-squashfs-sysupgrade.bin`

Signed-off-by: Qing W. <ses1er@gmail.com>
Tested-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>
Link: https://github.com/openwrt/openwrt/pull/20760
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

ses1er commented on target/linux/mediatek/dts/mt7986a-wavlink-wl-wn536ax6-a.dts:
```diff
@@ -0,0 +1,376 @@
+
+	aliases {
+		serial0 = &uart0;
+		label-mac-device = &wifi; 
```
removed, thanks!

ses1er commented on target/linux/mediatek/dts/mt7986a-wavlink-wl-wn536ax6-a.dts:
```diff
@@ -0,0 +1,376 @@
+		regulator-always-on;
+	};
+};
+
```
added and the following showed up:

```
[    7.345455] crypto-safexcel 10320000.crypto: EIP97:230(0,1,4,4)-HIA:270(0,5,5),PE:150/433(alg:7fcdfc00)/0/0/0
```

thank you!

xdarklight commented:
> > For users that want to backup their whole flash they need to add a custom partition that spans the whole 128MiB. Then with `CONFIG_PACKAGE_nand-utils=y` they can: `nanddump --file <directory>/full.nanddump /dev/mtd6` (assuming `mtd6` the partition that spans the whole flash)
> 
> Did you perform a full nand backup with adding a custom partition? can you please share the partition info?

yes, I did.
Similar to the BT Home Hub 5A install-image (which is only used for installation, so that partition is not part of OpenWrt by default) https://github.com/mkresin/BTHH5a-installimage/commit/76e1aba99184bc84fc955dd36d1c0a0ce942afb7#diff-997770ddcfe18a7cca809b3e0df94bb068521df6c023da91f79a6cafcc7f566fR64-R67 I added:
```
			partition@ff {
				label = "nand_chip";
				reg = <0x0 0x8000000>;
				read-only;
			};
```
as the last partition.
`partition@ff` is not correct obviously, I simply chose it not to conflict with `partition@0`

My dump shows that there's `0x3200000` bytes (50MiB) of free space (all `0xff`) after the end of `partition@7580000` (HW partition). The last `0x8000000` bytes (8MiB) seem to contain the "NMBM" (see `package/boot/uboot-mediatek/patches/100-04-mtd-add-core-facility-code-of-NMBM.patch`), so these must not be part of a partition as far as I understand.

xdarklight reviewed (approved):
`Tested-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>`

hauke commented:
It would be nice if OpenWrt would generate a image like the one you have at https://github.com/ses1er/firmware-misc/tree/main/wavlink/wl-wn536ax6a
What is needed to create this one?
